### PR TITLE
new image/image update resources

### PR DESCRIPTION
**Pull request by simao**
https://advancedtelematic.atlassian.net/browse/PRO-1717

### CODE REVIEW

**Comment by simao** [comment]
Also some other goodies (see individual commits)

File: core/src/test/scala/org/genivi/sota/core/image/ImageUpdateRepositorySpec.scala [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+      image <- imageRepository.persist(image.namespace, image.commit, image.ref, image.description, image.pullUri)
+      inserted <- imageUpdateRepository.persist(imageUpdate.namespace, image.id, imageUpdate.device)
+      updated <- imageUpdateRepository.persist(imageUpdate.namespace, image.id, imageUpdate.device)
+    } yield (inserted, updated)
+
+    val (inserted, updated) = f.futureValue
+
+    inserted.id shouldBe updated.id
+
+    inserted.updatedAt.isBefore(updated.updatedAt) shouldBe true
```
Comment (Danten) [inline core/src/test/scala/org/genivi/sota/core/image/ImageUpdateRepositorySpec.scala]: This seems to be what the next test, tests?

File: core/src/test/scala/org/genivi/sota/core/image/ImageResourceSpec.scala [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+  with ImageRepositorySupport {
+
+  implicit val _db = db
+
+  val route = new ImageResource(defaultNamespaceExtractor).route
+
+  test("POST Creates a new image") {
+    val commit: Commit = commitGenerator.generate
+    val refName = refNameGenerator.generate
+    val req = ImageRequest(commit, refName, "desc", Refined.unsafeApply("http//ats.com"))
```
Comment (Danten) [inline core/src/test/scala/org/genivi/sota/core/image/ImageResourceSpec.scala]: We might as well have http://ats.com

File: core/src/main/scala/org/genivi/sota/core/image/ImageUpdateResource.scala [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+  val ImageIdMatcher: PathMatcher1[ImageId] = JavaUUID.map(u => ImageId(u))
+
+  val DeviceIdMatcher: PathMatcher1[Uuid] = JavaUUID.map(Uuid.fromJava)
+
+  val imageUpdateProcess = new ImageUpdateProcess
+
+  val userRoutes = namespace { ns =>
+    path("image_update" / ImageIdMatcher / DeviceIdMatcher) { (image, device) =>
+      val f = imageUpdateProcess.queue(ns, image, device)
+      complete(f)
```
Comment (Danten) [inline core/src/main/scala/org/genivi/sota/core/image/ImageUpdateResource.scala]: Inline f here?

File: core/src/main/scala/org/genivi/sota/core/image/ImageUpdateResource.scala [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+    }
+  }
+
+  val deviceRoutes =
+    path("image_update" / DeviceIdMatcher) { device =>
+      complete(imageUpdateProcess.findPendingFor(device).map(_.toResponse))
+  }
+
+  val route = handleErrors { userRoutes ~ deviceRoutes }
```
Comment (Danten) [inline core/src/main/scala/org/genivi/sota/core/image/ImageUpdateResource.scala]: Should deviceRoutes maybe have a separate prefix from the userRoutes? In core the device routes is under the mydevice prefix.

Comment (simao) [inline core/src/main/scala/org/genivi/sota/core/image/ImageUpdateResource.scala]: The thing is the ui may at the moment use this route as well... We can keep it for now and separate them when we need two different effects...

**Comment by simao** [comment]
The reason the data types are separated is because I wanted to keep everything as contained as possible so it's easier to extract later, I am not sure this couldn't be extracted to a separate service for example.

Also, I would prefer if other services didn't use these data types, at least for the time being, because they are very unstable until we figure out everything we need.

File: core/src/test/scala/org/genivi/sota/core/image/ImageUpdateRepositorySpec.scala (hunk 1 above)
Comment (simao) [inline core/src/test/scala/org/genivi/sota/core/image/ImageUpdateRepositorySpec.scala]: yep

File: core/src/main/scala/org/genivi/sota/core/image/ImageResource.scala [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+  val userRoutes =
+    path("image_update" / ImageIdMatcher / DeviceIdMatcher) { (image, device) =>
+      complete("")
+    }
+
+  val deviceRoutes =
+    path("image_update" / DeviceIdMatcher) { device =>
+      complete("")
+  }
+
```
Comment (CalumMcCall) [inline core/src/main/scala/org/genivi/sota/core/image/ImageResource.scala]: indentation?

File: core/src/main/scala/org/genivi/sota/core/image/ImageResource.scala [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+  extends Directives with ImageRepositorySupport {
+
+  import ImageResource._
+
+  val route = namespace { ns =>
+    path("image") {
+      post {
+        entity(as[ImageRequest]) { img =>
+          val f = imageRepository.persist(Image(ns, ImageId.generate(), img.commit, img.ref, img.description))
+          complete(f)
```
Comment (CalumMcCall) [inline core/src/main/scala/org/genivi/sota/core/image/ImageResource.scala]: inline the f?

File: core/src/test/scala/org/genivi/sota/core/image/ImageUpdateResourceSpec.scala [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+  with ImageGenerators
+  with DefaultPatience
+  with ImageRepositorySupport {
+
+  implicit val _db = db
+
+  val route = new ImageUpdateResource(defaultNamespaceExtractor).route
+
+  test("fail") {
+    fail("wat")
```
Comment (CalumMcCall) [inline core/src/test/scala/org/genivi/sota/core/image/ImageUpdateResourceSpec.scala]: wat?

Comment (simao) [inline core/src/test/scala/org/genivi/sota/core/image/ImageUpdateResourceSpec.scala]: ahah yeah this was removed in a subsequent commit

**Pull request merged**